### ISSUE TITLE
Distribution factory functions

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#22427]
This issue discusses the set of factory functions to be provided by a distribution that is to be considered stable for Chapel 2.0. It is a part of our short-term strategy to address #17908.

The proposed factory functions and their names+signatures are as follows.

Legend:
`dist.` is a distribution value; distribution values are parallel to instances of domain and array types
`RD` is short for "a domain argument or range(s) argument(s)"
`args...` is 0 or more distribution-specific arguments
`Block` is a stand-in for the distribution name

| purpose | name |
| :--- | :-- |
| create a Block dist. | proc type Block.createDistribution(args...) |
| create a Block-distributed domain | proc type Block.createDomain(RD, args...) |
| create a Block-distributed array | proc type Block.createArray(RD, eltType, args...) |
| create a domain for the given Block dist. | proc Block.createDomain(RD) |
| is this dist/dom/array Block-distributed? | proc (dist.\|domain\|array).isBlock() |

**Comment by lydia-duncan** [comment]
Because #22219 was able to be closed, I think this one can be as well

**Comment by vasslitvinov** [comment]
@bradcray what is the stable way to provide this functionality?
```chpl
proc Block.createDomain(domain or range(s))
```

**Comment by bradcray** [comment]
Actually, I don't think this one can be closed, though I do think the 2.0 label can be removed.  Specifically, the second-to-last row of the table doesn't exist today (note that it's a value, rather than type, method), and would definitely be useful—specifically, it'd be the only currently stable way to create a number of domains sharing a single distribution object (at least, unless/until we come up with a stable replacement for `dmapped`—Vass, did the subteam you convened not agree on `distributed` for that purpose?  I can't recall and haven't gone looking for the notes.

@jeremiah-corrado said something to me suggesting that he might pick that item up, along with some other factory improvements he had in mind.

**Comment by jeremiah-corrado** [comment]
Yeah, I can add the value-method to the PR I'm working that improves the type-methods.

**Comment by vasslitvinov** [comment]
> at least, unless/until we come up with a stable replacement for dmapped—Vass, did the subteam you convened not agree on distributed for that purpose? 

The subteam concluded, among others:
* to use `(a domain).distributed` as a replacement for `(a domain).dist`, not for `dmapped`,
* to use factory functions as a stable replacement for the now-unstable `dmapped`, and
* the specifics of those factory functions were to be discussed on this issue.

**Comment by jeremiah-corrado** [comment]
Thanks @vasslitvinov, I've started implementing [changes to the factory functions](https://github.com/chapel-lang/chapel/pull/23052/files) on `Block`, `Cyclic` and `Stencil`, but I did not realize until now that those were going to be considered part of Chapel 2.0.

This is the interface for `Block` that I currently have:
```chapel
proc type createDomain(dom: domain, targetLocales = Locales);
proc type createDomain(rng: range..., targetLocales = Locales);
proc type createArray(dom: domain, type eltType, targetLocales = Locales);
proc type createArray(rng: range..., type eltType, targetLocales = Locales);
proc type createArray(dom: domain, type eltType, value: eltType, targetLocales = Locales);
proc type createArray(rng: range..., type eltType, value: eltType, targetLocales = Locales);
```

The `Cyclic` and `Stencil` interfaces are essentially the same except that `Stencil` also has optional arguments for `fluff` and `periodic`. 

Does this agree with what was decided in the ad-hoc subteam on this issue?

**Comment by bradcray** [comment]
Following up on Vass's comment, in the perf meeting a few weeks ago, we discussed having the createArray() routines take three types of initialing expressions:
* a scalar of the array's type
* an array of similar eltType
* an iterable expression (`_iteratorRecord`)

I think this could be done easily with three overloads, or with a single generic routine that filtered for these cases.  I'd probably call the argument something like `initExpr` or `initializer` since it's not a singleton value.

One thing that worries me slightly is that I believe BHarsh has been telling me lately that arguments following varargs don't work very well (unless I've mis-internalized the case).  So I'd check that in:

```chapel
proc type createDomain(rng: range..., targetLocales = Locales);
```

you can pass targetLocales to it.  (I definitely think this is something we want the language to support, where a named argument would likely be required to stop the varargs matching, but Ben wanted to use it for optional serializers in writeln() and couldn't get it to work.  Maybe that's specific to varargs without a declared type, though?

Also, note that `range` should probably be `range(?)` to accept ranges of varying idxTypes, strideKinds, etc.

**Comment by jeremiah-corrado** [comment]
> One thing that worries me slightly is that I believe BHarsh has been telling me lately that arguments following varargs don't work very well (unless I've mis-internalized the case).

This is a challenge, but I think I have it working by providing multiple overloads of the methods that accept ranges (one with a targetLocales argument, and one without).

**Comment by jeremiah-corrado** [comment]
> * an array of similar eltType

For this version, it seems like it would make more sense to be able to write:
```chapel
var myDistArray = Block.createArray(myArray);
```
Since the domain and element type are already features of `myArray`. 

Otherwise, I think calls might end up looking like:
```chapel
var myDistArray = Block.createArray(myArray.domain, myArray.eltType, myArray);
```

**Comment by bradcray** [comment]
Possibly...  This is why we were discussing having a distinct `tryCopyArray()` call in the perf meeting the other week, while still permitting something like "I'd like to declare a distributed array of reals that is implemented by this non-distributed array (so it has a different domain) of integers (and a different element type).

Another option would be to put the `initExpr=` argument first, and then have the domain and eltType arguments support defaults based on it.  But then we've got the issue of different overloads having different arguments in different orders which seems... a little unfortunate.

**Comment by jeremiah-corrado** [comment]
https://github.com/chapel-lang/chapel/pull/23112 partially implements the interface described above. The `createDistribution` and `isBlock` methods have not been implemented yet. Additionally, these methods could be added to more of the distribution types (where they are currently supported on `Block`, `Cyclic` and `Stencil`).

#23112 also includes overloads of `createArray` that can accept a value, an iterator expression, or another array. These methods have been marked unstable for now as the whole team was not on board with stabilizing them during a recent team meeting. Particularly, the following concern was voiced again:

> For this version, it seems like it would make more sense to be able to write:
> ```chapel
> var myDistArray = Block.createArray(myArray);
> ```
> Since the domain and element type are already features of `myArray`. 
>
> Otherwise, I think calls might end up looking like:
> ```chapel
> var myDistArray = Block.createArray(myArray.domain, myArray.eltType, myArray);
> ```

**Comment by bradcray** [comment]
I don't think createDistribution is necessary now that distributions are records—it duplicates the initializer logic.

I don't feel confident that `isBlock` got any design discussion or consensus, but may just be misremembering.

With respect to this opinion:

> For this version, it seems like it would make more sense to be able to write:
> ```chapel
> var myDistArray = Block.createArray(myArray);
> ```
> Since the domain and element type are already features of `myArray`. 
>
> Otherwise, I think calls might end up looking like:
> ```chapel
> var myDistArray = Block.createArray(myArray.domain, myArray.eltType, myArray);
> ```

I think we should support this with a different method, `copyArray(myArray)`.  That permits the `createArray()` call to be  more general/powerful (by permitting changes to index set or element type) and having a simpler/more direct way to do a copy.  It seems like the main alternative would be to define:

```chapel
Block.createArray(initExpr: [?D] ?et, dom: domain(?) = D, type eltType=et);
```

which is attractive in that it permits `Block.createArray(myArray);` but unattractive in that it changes the argument order relative to all of the other variations (so requires a user to keep all of the possible signatures for `createArray()` straight).
